docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/EventEmitter.java

```
@@ -0,0 +1,162 @@
+            }
+        };
+
+        return cluster -> emitNotThrowing(
+                () -> recordingClusterWait.accept(cluster),
+                task -> Event.clusterWait(ClusterWaitEvent.builder()
+                        .task(task)
+                        .serviceNames(recordedServiceNames.get().orElseThrow(
+                                () -> new IllegalStateException("Recorded service names have not yet been computed")))
+                        .type(clusterWaitType)
+                        .build()));
+    }
+
+    private void emitNotThrowing(CheckedRunnable runnable, Function<Task, Event> eventFunction) {
+        try {
+            emitTask(runnable, eventFunction);
+        } catch (InterruptedException | IOException e) {
+            Throwables.propagate(e);
+        }
+    }
+
+    private void emitTask(CheckedRunnable runnable, Function<Task, Event> eventFunction)
+            throws IOException, InterruptedException {
+        Optional<String> failure = Optional.empty();
+        OffsetDateTime startTime = clock.instant().atOffset(ZoneOffset.UTC);
+        try {
+            runnable.run();
+        } catch (RuntimeException | IOException | InterruptedException e) {
+            failure = Optional.of(Exceptions.condensedStacktraceFor(e));
+            throw e;
+        } finally {
+            OffsetDateTime endTime = clock.instant().atOffset(ZoneOffset.UTC);
+            Task task = Task.builder()
+                    .startTime(startTime)
+                    .endTime(endTime)
+                    .failure(failure)
+                    .build();
+
+            Event event = eventFunction.apply(task);
+            emitEvent(event);
+        }
+    }
+
+    private void emitEvent(Event event) {
+        eventConsumers.forEach(eventConsumer -> {
+            try {
+                eventConsumer.receiveEvent(event);
+            } catch (Exception e) {
+                log.error("Error sending event {}", event, e);
```

I think we should propagate the error if an eventConsumer fails

---

What happens if you have multiple event consumers? Then if one fails to consume an event the rest don't get the event.

---

I'm trying to avoid unexpected behaviour when a user provides their own consumers. Its kinda weird, but I could imagine users adding event consumers to ensure some sort of preconditions, and failing the test (suite|case) if they do not hold

---

Ok, this will now throw and add all exceptions that happened as suppressed.